ambry-replication/src/main/java/com/github/ambry/replication/ReplicationEngine.java

```diff
@@ -503,4 +507,23 @@ protected void stopPartitionReplication(PartitionId partitionId) {
       }
     }
   }
+
+  /**
+   * Go through remote replicas for this partition and compare messages written to local store with the missing messages
+   * found during previous meta data exchange. If there are matching messages (based on store key), remove them from the missing message set.
+   * This is used during leader-based replication to update token for standby replicas. Standby replicas store the
+   * missing messages in metadata exchange, track them through intra-dc replication and update token when all the
+   * missing messages are written to store.
+   * @param partitionId partition ID of the messages written to store
+   * @param messageInfoList list of messages written to store
+   */
+  void onMessageWriteForPartition(PartitionId partitionId, List<MessageInfo> messageInfoList) {
+    rwLock.readLock().lock();
+    try {
+      PartitionInfo partitionInfo = partitionToPartitionInfo.get(partitionId);
+      partitionInfo.updateReplicaInfosOnMessageWrite(messageInfoList);
```

Comment:
You need to check if `partitionInfo == null` here,  it's possible partition is being concurrently removed from this node